Initial REST sample

jt-nti opened the pull request:
Needs more unhappy path testing and lots of review comments!

mbwhite commented:
Having spoken to @jt-nti here are a summary of the comments I had working with this PR (it did work btw).  It primarily focuses on documentation/code comments.

We need  to ensure that the code is well signposted so it's easy to tell the difference between the boilerplate REST code, the Fabric code, how the configuration is done.

denyeart commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+
+HLF_PRIVATE_KEY_ORG2="$(cat ${PRIVATE_KEY_FILE_ORG2} | sed -e 's/$/\\n/' | tr -d '\r\n')"
+
+HLF_COMMIT_TIMEOUT=3000
+
+HLF_ENDORSE_TIMEOUT=30
```
Are these in different units?
Good to put the units as a comment to make it clear.

jt-nti replied:
I thought they were the same unit but I'll check

generateEnv.sh is a convenience script for running the sample and the units are documented in config.ts - I'd prefer to avoid duplicating the comments but it would definitely help to have a comment at the top of the generated file pointing to config.ts, which I think relates to one of Matthew's comments

jt-nti replied:
They are both seconds
hyperledgendary/fabric-rest-sample#54

denyeart commented on asset-transfer-basic/rest-api-typescript/src/__tests__/api.test.ts:
```diff
@@ -0,0 +1,738 @@
+/*
```
Are there instructions somewhere about how to generate the test fixtures?

jt-nti replied:
I'm not sure what you mean- there aren't any generated test fixtures.

denyeart commented on asset-transfer-basic/rest-api-typescript/src/assets.router.ts:
```diff
@@ -0,0 +1,348 @@
+      });
+    }
+
+    const mspId = req.user as string;
```
mspid is for the org. isn't req.user an individual user?

jt-nti replied:
There's currently an API for each org, rather than an individual user

denyeart commented on asset-transfer-basic/rest-api-typescript/src/config.ts:
```diff
@@ -0,0 +1,274 @@
+ */
+export const commitTimeout = env
+  .get('HLF_COMMIT_TIMEOUT')
+  .default('3000')
```
3000 seconds, really?

jt-nti replied:
That does seem suspicious! I'll check!

jt-nti replied:
Updated to match the SDK default of 300 seconds
hyperledgendary/fabric-rest-sample#54

denyeart commented on asset-transfer-basic/rest-api-typescript/src/config.ts:
```diff
@@ -0,0 +1,274 @@
+/*
+ * Certificate for the Org1 identity
+ */
+export const certificateOrg1 = env
```
certificateOrg1User1

denyeart commented on asset-transfer-basic/rest-api-typescript/src/config.ts:
```diff
@@ -0,0 +1,274 @@
+/*
+ * Certificate for the Org2 identity
+ */
+export const certificateOrg2 = env
```
certificateOrg2User1

denyeart commented on asset-transfer-basic/rest-api-typescript/src/errors.ts:
```diff
@@ -0,0 +1,218 @@
+/*
+ * Handles errors from evaluating and submitting transactions.
+ *
+ * As with duplicate transaction errors, checking error strings like this is
+ * not ideal. Unfortunately the chaincode samples do not use error codes so
+ * again it's the only option. The error message text is not even the same for
+ * the Go, Java, and Javascript implementations of the chaincode!
+ */
```
Let's see if we can resolve this in v2.4.

jt-nti replied:
There's a limit to what's possible with current chaincode since you only ever get back a message but I'll open an issue and we can have another chat about errors

jt-nti replied:
Finally got round to opening that issue! https://github.com/hyperledger/fabric-samples/issues/543

denyeart commented on asset-transfer-basic/rest-api-typescript/src/fabric.ts:
```diff
@@ -0,0 +1,325 @@
+        'Fatal transaction error occurred'
+      );
+
+      // Return a job result to stop retrying
+      return {
+        transactionError: err.toString(),
+        transactionPayload: undefined,
+      };
```
It is not clear how returning a job result will stop retries.
Also, transactionError and transactionPayload does not intuitively convey a job result.

denyeart reviewed (commented):
For my second pass, I focused on the retry logic which was one of the core concepts we wanted to demonstrate with this sample. My comments in fabric.ts suggest that the user will have a difficult time understanding how the retries work. I had to read that part of the code many times, and then find and understand all the aspects of job submission and retries in other files, before I could piece it all together and understand how it all works together. Somebody coming into this code without any context would have a more difficult time with it.

Taking a step back, my larger feedback is that the use of BullMQ backed by Redis are too heavyweight for what we are trying to accomplish in this sample. They get in the way of the core concepts that we are trying to convey around retries.  I was envisioning a code block that could stand by itself to demonstrate the retries. Rather than persisted queued jobs, I think we can simply keep the requests in memory and add a few comments stating that the requests _could_ be backed by message queues if you want persisted jobs. The implementer will probably want their own favorite queuing/backend system regardless, rather than using BullMQ/Redis.

If you really think we should demonstrate the BullMQ/Redis stuff, then we will need to find a way to make the core concepts and implementation easier to understand for a newbie.

jt-nti commented on .github/workflows/publish.yaml:
```diff
@@ -0,0 +1,54 @@
+      - name: Build image
+        run: docker build --tag $IMAGE_NAME --label "runnumber=${GITHUB_RUN_ID}" $SOURCE_FOLDER 
+
+      - name: Log in to registry
+        # This is where you will update the PAT to GITHUB_TOKEN
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login ghcr.io -u ${{ github.actor }} --password-stdin
+
+      - name: Push image
+        run: |
+          IMAGE_ID=ghcr.io/${{ github.repository_owner }}/$IMAGE_NAME
```
The image is pushed for @jkneubuh's k8s work in the [REST samples initial resting place](https://github.com/hyperledgendary/fabric-rest-sample) and it seems reasonable to push to hyperledger/fabric-samples when it's merged here

jt-nti commented on asset-transfer-basic/rest-api-typescript/README.md:
```diff
@@ -0,0 +1,184 @@
+
+## Usage
+
+**Note:** these instructions should work with the release-2.2 branch of `fabric-samples` but later versions require some changes
```
Just a point in time thing as 2.2 was the most stable- updating to 2.4 with the Gateway is definitely on the list

jt-nti replied:
Checked it works with the 2.4 release and main branch of fabric-samples
hyperledgendary/fabric-rest-sample#51

jt-nti commented on asset-transfer-basic/rest-api-typescript/README.md:
````diff
@@ -0,0 +1,184 @@
+TEST_NETWORK_HOME=$HOME/fabric-samples/test-network npm run generateEnv
+```
+
+Start a Redis server
````
See second pass comment

jt-nti commented on asset-transfer-basic/rest-api-typescript/package-lock.json:
```diff
@@ -0,0 +1,6830 @@
+{
```
The is an application rather than a library so I think checking in package-lock.json is the right thing to do

denyeart replied:
For every other application sample, we have decided not to check in package-lock.json since it gets auto-generated for people with the latest dependencies.

jt-nti replied:
That doesn't seem like a good example to set, and the rest sample uses `npm ci` in the docker image

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+
+: "${TEST_NETWORK_HOME:=../..}"
+: "${CONNECTION_PROFILE_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/connection-org1.json}"
+: "${CERTIFICATE_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/signcerts/Admin@org1.example.com-cert.pem}"
```
I'll have a try

jt-nti replied:
The User1 credentials work, so I've switched those over hyperledgendary/fabric-rest-sample#53
I'll leave the variable names as they are since they are to configure any suitable org1 user, not necessarily User1

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+: "${TEST_NETWORK_HOME:=../..}"
+: "${CONNECTION_PROFILE_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/connection-org1.json}"
+: "${CERTIFICATE_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/signcerts/Admin@org1.example.com-cert.pem}"
+: "${PRIVATE_KEY_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/keystore/priv_sk}"
```
same :)

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+: "${PRIVATE_KEY_FILE_ORG1:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp/keystore/priv_sk}"
+
+: "${CONNECTION_PROFILE_FILE_ORG2:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org2.example.com/connection-org2.json}"
+: "${CERTIFICATE_FILE_ORG2:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org2.example.com/users/Admin@org2.example.com/msp/signcerts/Admin@org2.example.com-cert.pem}"
```
same :)

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+
+: "${CONNECTION_PROFILE_FILE_ORG2:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org2.example.com/connection-org2.json}"
+: "${CERTIFICATE_FILE_ORG2:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org2.example.com/users/Admin@org2.example.com/msp/signcerts/Admin@org2.example.com-cert.pem}"
+: "${PRIVATE_KEY_FILE_ORG2:=${TEST_NETWORK_HOME}/organizations/peerOrganizations/org2.example.com/users/Admin@org2.example.com/msp/keystore/priv_sk}"
```
same :)

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+
+MAX_RETRY_COUNT=5
+
+HLF_CERTIFICATE_ORG1="$(cat ${CERTIFICATE_FILE_ORG1} | sed -e 's/$/\\n/' | tr -d '\r\n')"
```
same :)

jt-nti commented:
@denyeart I've been scratching my head trying to make the retry logic clearer but obviously it still needs some work. This is meant to be a production sample and after the last group chat that message was reinforced, so something like Redis and BullMQ are definitely required- there are other options but those seemed the most lightweight and a good fit for a TypeScript sample. BullMQ will retry a job if it fails (throws an error) and won't otherwise, which is pretty straightforward, however we have the usual fun of happy errors and unhappy errors to deal with. I'll have another think about how best to separate out the BullMQ and Fabric logic a bit more clearly.

jt-nti commented on asset-transfer-basic/rest-api-typescript/scripts/generateEnv.sh:
```diff
@@ -0,0 +1,76 @@
+
+HLF_CERTIFICATE_ORG1="$(cat ${CERTIFICATE_FILE_ORG1} | sed -e 's/$/\\n/' | tr -d '\r\n')"
+
+HLF_PRIVATE_KEY_ORG1="$(cat ${PRIVATE_KEY_FILE_ORG1} | sed -e 's/$/\\n/' | tr -d '\r\n')"
```
The REST app loads all its config from environment variables- the generateEnv.sh script is a helper to create a .env file to make running locally easier

denyeart commented:
> @denyeart I've been scratching my head trying to make the retry logic clearer but obviously it still needs some work. This is meant to be a production sample and after the last group chat that message was reinforced, so something like Redis and BullMQ are definitely required- there are other options but those seemed the most lightweight and a good fit for a TypeScript sample. BullMQ will retry a job if it fails (throws an error) and won't otherwise, which is pretty straightforward, however we have the usual fun of happy errors and unhappy errors to deal with. I'll have another think about how best to separate out the BullMQ and Fabric logic a bit more clearly.

Ok, I agree a production system would use some type of queue/persistence for submitted transactions. But then again a production system using this pattern would likely have the client submit directly via queue rather than REST so we are already deviating from what a production system would do. Ultimately I think we need to prioritize making the retry pattern extremely readable and clear to the reader... if that can be done with a queue approach, great.

Maybe @mbwhite or @bestbeforetoday have ideas...

jt-nti commented on asset-transfer-basic/rest-api-typescript/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+FROM node:14-alpine3.12 AS build
```
Updated hyperledgendary/fabric-rest-sample#52

jt-nti commented:
> Having spoken to @jt-nti here are a summary of the comments I had working with this PR (it did work btw). It primarily focuses on documentation/code comments.
> 
> We need to ensure that the code is well signposted so it's easy to tell the difference between the boilerplate REST code, the Fabric code, how the configuration is done.

Added a bit more signposting for how the configuration is done hyperledgendary/fabric-rest-sample#54

jt-nti commented:
> I was envisioning a code block that could stand by itself to demonstrate the retries.

I've moved the job queue implementation details out of the fabric.ts file completely, and added the following retry logic to errors.ts to hopefully clarify what's going on

```
/*
 * Get the required transaction retry action for an error.
 *
 * For this sample transactions are considered retriable if they fail with any
 * error, *except* for duplicate transaction errors, or errors from the smart
 * contract.
 *
 * You might decide to retry transactions which fail with specific errors
 * instead, for example:
 *   MVCC_READ_CONFLICT
 *   PHANTOM_READ_CONFLICT
 *   ENDORSEMENT_POLICY_FAILURE
 *   CHAINCODE_VERSION_CONFLICT
 *   EXPIRED_CHAINCODE
 */
export const getRetryAction = (err: unknown): RetryAction => {
  if (isDuplicateTransactionError(err) || err instanceof ContractError) {
    return RetryAction.None;
  } else if (err instanceof TimeoutError) {
    return RetryAction.WithExistingTransactionId;
  }

  return RetryAction.WithNewTransactionId;
};
```

mbwhite commented on asset-transfer-basic/rest-api-typescript/.env.sample:
```diff
@@ -0,0 +1,39 @@
+LOG_LEVEL=info
```
Worth adding comments to this file?

jt-nti replied:
Is this ok? https://github.com/hyperledgendary/fabric-rest-sample/blob/main/asset-transfer-basic/rest-api-typescript/.env.sample#L1-L7

mbwhite commented on asset-transfer-basic/rest-api-typescript/README.md:
```diff
@@ -0,0 +1,184 @@
+
+## Usage
+
+**Note:** these instructions should work with the release-2.2 branch of `fabric-samples` but later versions require some changes
```
but this is being merged into the main branch - this doesn't make sense...

jt-nti replied:
Updated here https://github.com/hyperledgendary/fabric-rest-sample/blob/main/asset-transfer-basic/rest-api-typescript/README.md?plain=1#L30

mbwhite commented on asset-transfer-basic/rest-api-typescript/README.md:
````diff
@@ -0,0 +1,184 @@
+TEST_NETWORK_HOME=$HOME/fabric-samples/test-network npm run generateEnv
+```
+
+Start a Redis server
````
Is the purpose of the Redis image explained somewhere?

jt-nti replied:
Possibly, but I'll add something here as well!

jt-nti replied:
https://github.com/hyperledgendary/fabric-rest-sample/pull/55/files#diff-6f105b2df9bc613cdff4936b24a577d587b5aa784c9400ac2f1f0967d3696ed5R54

mbwhite commented on asset-transfer-basic/rest-api-typescript/README.md:
```diff
@@ -0,0 +1,184 @@
+
+## REST API
+
+If everything went well, you can now open a new terminal and try out some basic asset transfer REST calls!
```
Should the `ready` and the `live` APIs be mentioned here...

jt-nti replied:
I'm leaning towards keeping it simple and not mentioning them in the readme. Seems like too much of a tangent into hosting for a high level run through.

mbwhite commented on asset-transfer-basic/rest-api-typescript/src/redis.ts:
```diff
@@ -0,0 +1,51 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * TBC
```
TBC...

jt-nti replied:
Should probably write something more useful there!!

jt-nti replied:
https://github.com/hyperledgendary/fabric-rest-sample/pull/55/files#diff-19955bdd936aa9c7f85937db6ec2aede745e91e8b4de72bb2f6f09a0ea173149R4

kingworm commented on asset-transfer-basic/rest-api-typescript/package.json:
```diff
@@ -0,0 +1,63 @@
+{
+  "name": "asset-transfer-basic",
```
1. I suggest you to specify node, and npm version requirements on this package file.
2. I see that npm version is quite low ( `lockfileVersion` is 1 )

jt-nti replied:
Hi @kingworm, thanks, I'll add that

kingworm commented on asset-transfer-basic/rest-api-typescript/tsconfig.json:
```diff
@@ -0,0 +1,75 @@
+{
+  "compilerOptions": {
```
I think you can remove this commented options

mbwhite reviewed (approved):
Good starting point we can build from for a showing production patterns.
